### PR TITLE
Adding Conformer encoder I/O-styled Transformer encoder

### PR DESCRIPTION
**Pull request by tango4j**
What does this PR do?

Follow up work after the initial TF encoder PR (https://github.com/NVIDIA-NeMo/NeMo/pull/15661). Many NeMo Speech AI maintainers are asking for the new Transformer Encoder implementations to have pre-encode, positional encoding features in conformer encoder.

Aligns the ASR `TransformerEncoder` module with the offline `ConformerEncoder` module surface while preserving Transformer-specific attention parameters and behavior.

Streaming encoder and adapter implementations are not included in this PR. These features will be added later on.

Tested LibriSpeech training with Transformer + CTC (BPE). Added `transformer_ctc_bpe.yaml` with the default configurations.

**Collection**: ASR

# Changelog

- Updated `TransformerEncoder` to inherit NeMo module/export/access mixins and expose Conformer-style input/output type metadata.
- Added Conformer-style offline encoder utilities, including `input_example`, `forward_for_export`, `forward_internal`, `bypass_pre_encode`, `feat_out`, positional encoding, pad mask toggling, stochastic depth, and inter-CTC tensor capture.
- Added Conformer-style pre-encoder options while preserving the Transformer-native `FeatureStacking` path as `subsampling="feature_stacking"`.
- Moved `FeatureStacking` into the shared ASR subsampling module so it can be imported from `nemo.collections.asr.parts.submodules.subsampling`.
- Added `self_attention_model` mirroring Conformer's positional-encoding switch: `"rel_pos"` (default), `"abs_pos"`, and `"no_pos"` (`None` is accepted as a YAML alias for `"no_pos"`).
- Implemented Transformer-XL relative PE on FlexAttention — (b)+(d) bias via a `score_mod` closure, (c) bias folded as `Q + pos_bias_u`; rel-shift is shared with `ConformerEncoder` via `RelPositionMultiHeadAttention.rel_shift`.
- Added Transformer encoder tests mirroring relevant Conformer encoder test procedures for stochastic depth and bypass pre-encode behavior.
- Added `self_attention_model` tests, including a `T != n_heads` regression for `pos_bias_{u,v}` broadcasting.
- Updated Transformer encoder tests to use typed-module keyword arguments and validate output lengths.
- Wrapped CPU forward tests in `torch.no_grad()` so FlexAttention's CPU path doesn't raise under `model.train()`.

# Usage

```python
from nemo.collections.asr.modules.transformer_encoder import TransformerEncoder

encoder = TransformerEncoder(
    feat_in=128,
    d_model=512,
    n_heads=8,
    n_layers=17,
    subsampling="feature_stacking",
    subsampling_factor=4,
    self_attention_model="rel_pos",  # one of "rel_pos" | "abs_pos" | "no_pos" (or None)
)

encoded, encoded_len = encoder(audio_signal=features, length=feature_lengths)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

Follow up work after the initial TF encoder PR (https://github.com/NVIDIA-NeMo/NeMo/pull/15661)

### CODE REVIEW

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 1]
```diff
@@ -165,37 +135,118 @@ class TransformerEncoder(nn.Module):
...
         drop_rate: float = 0.1,
+        dropout_pre_encoder: float = None,
+        dropout_emb: float = 0.0,
         qkv_bias: bool = False,
         qk_norm: bool = False,
         ff_expansion: float = 4.0,
         pre_block_norm: bool = True,
-        subsampling_factor: int = 4,
+        pos_emb_max_len: int = 5000,
+        xscaling: bool = True,
```
Comment (stevehuang52) [inline nemo/collections/asr/modules/transformer_encoder.py]: Shall we set default xscaling to False, since we already know that the layernorm will zero-out the effect of xscaling?

Comment (tango4j) [reply nemo/collections/asr/modules/transformer_encoder.py]: Thanks for pointing this out. Setting this with default to False.

**Comment by tango4j** [comment]
/ok to test [5398604](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/539860402976a7d094e0f8c4b8ccabd01cd6da1f)

**Comment by nithinraok** [comment]
Thanks Taejin!

Have you had a chance to run training with this? Does it converges similarly with positional embedding enabled and how the results compare to previous runs.

**Comment by tango4j** [comment]
> Thanks Taejin!
> 
> Have you had a chance to run training with this? Does it converges similarly with positional embedding enabled and how the results compare to previous runs.

@KunalDhawan is working on using this PR part to train his MoE transformer experiments.
If time allows, I will also try Librispeech-only training with several setups to do sanity check (to test the points you mentioned)

Recently, after doing some survey, It appeared to me that convnet frontend and positional encoding can affect the performance a lot. So I think we need to test these two configurations separately (ablations).

**Comment by tango4j** [comment]
@nithinraok 
I have tested `relative positional encoding` with training job. Compared with FastConformer CTC vs Transformer CTC.
Having `rel_pos` in the Transformer encoder affects the performance a lot (10~20% more error). Need to set this as default.

Also figured that Filterbank Stacking feature is equally good as dw_striding (3 level convnet frontend). Better switching to Filterbank stacking to make this model low precision friendly.

@ipmedenn @KunalDhawan @stevehuang52 
Now, transformer is fully equipped with relative positional encoding.

**Comment by tango4j** [comment]
/ok to test [6725930](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/67259308afc6c5aa35f729e4e774f2cb822741e5)

File: examples/asr/conf/transformer/transformer_ctc_bpe.yaml
Comment (nithinraok) [inline examples/asr/conf/transformer/transformer_ctc_bpe.yaml:1]: could you also move `https://github.com/NVIDIA-NeMo/NeMo/blob/main/examples/asr/conf/fastconformer/transformer_stacking_tdt_bpe.yaml` to this folder. `examples/asr/conf/transformer/`

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 2]
```diff
@@ -12,18 +12,77 @@
...
+    RelPositionalEncoding,
+    RelPositionMultiHeadAttention,
+)
+from nemo.collections.asr.parts.submodules.subsampling import ConvSubsampling, FeatureStacking, StackingSubsampling
+from nemo.collections.asr.parts.utils.regularization_utils import compute_stochastic_depth_drop_probs
+from nemo.core.classes.common import typecheck
+from nemo.core.classes.exportable import Exportable
+from nemo.core.classes.mixins import AccessMixin
+from nemo.core.classes.module import NeuralModule
+from nemo.core.neural_types import AcousticEncodedRepresentation, BoolType, LengthsType, NeuralType, SpectrogramType
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py]: Please remove support for all these

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 3]
```diff
@@ -12,18 +12,77 @@
...
 import torch
 import torch.nn as nn
 from torch.nn.attention.flex_attention import create_block_mask, flex_attention
 
+from nemo.collections.asr.parts.submodules.multi_head_attention import (
+    PositionalEncoding,
+    RelPositionalEncoding,
+    RelPositionMultiHeadAttention,
+)
+from nemo.collections.asr.parts.submodules.subsampling import ConvSubsampling, FeatureStacking, StackingSubsampling
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py]: Let's drop `ConvSubsampling` from this PR.

This module lives under ASR and profiling shows it underperforms `FeatureStacking` on training throughput and complicates inference-tool support, with no accuracy upside. Removing it keeps the PR focused and avoids carrying forward a suboptimal path. Users can add this later if they found it useful for other tasks

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: examples/asr/conf/transformer/transformer_ctc_bpe.yaml
Comment (tango4j) [reply examples/asr/conf/transformer/transformer_ctc_bpe.yaml:1]: moved the file

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 2 above)
Comment (tango4j) [reply nemo/collections/asr/modules/transformer_encoder.py]: removed all 6 requested lines.

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/modules/transformer_encoder.py]: Removed ConvSubsampling.

**Comment by tango4j** [comment]
/ok to test [317a59e](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/317a59ee51af616b24bb3fc914bf6836956fae6d)

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 4]
```diff
@@ -146,7 +236,20 @@ def forward(self, x, block_mask=None):
             q = self.q_norm(q).to(v.dtype)
             k = self.k_norm(k).to(v.dtype)
 
-        out = flex_attention_compiled(q, k, v, block_mask=block_mask)
+        score_mod = None
+        if self.self_attention_model == "rel_pos":
```
Comment (pzelasko) [inline nemo/collections/asr/modules/transformer_encoder.py]: all of these assertions can be executed in `__init__` instead of in `forward`, which will improve readability and execute once instead of every forward

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 5]
```diff
@@ -170,52 +273,118 @@ def forward(self, x, block_mask=None):
...
+        if self_attention_model is None:
+            self_attention_model = "no_pos"
+        if self_attention_model not in ("abs_pos", "rel_pos", "no_pos"):
+            raise ValueError(
+                f"self_attention_model='{self_attention_model}' is not supported. "
+                "Currently only 'abs_pos', 'rel_pos', and 'no_pos' (or None) are available."
+            )
+        if dropout_pre_encoder is None:
+            dropout_pre_encoder = drop_rate
+        if subsampling == 'feature-stacking':
```
Comment (pzelasko) [inline nemo/collections/asr/modules/transformer_encoder.py]: why are we adding backward-compatibility string remapping to a brand new class?

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 4 above)
Comment (tango4j) [reply nemo/collections/asr/modules/transformer_encoder.py]: made a variable "self._uses_rel_pos" to deal everything in __init__.

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 5 above)
Comment (tango4j) [reply nemo/collections/asr/modules/transformer_encoder.py]: These lines are removed.

**Comment by tango4j** [comment]
/ok to test [d7def7a](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/d7def7adead9189f6a6d08699b0427da9136182e)

**Comment by tango4j** [comment]
/ok to test [56423ba](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/56423ba7c3360870ed14d09d704ba820cb63dca3)

**Comment by tango4j** [comment]
/ok to test [839fd85](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/839fd85d63dd4758b065d3bc26a1acf44ac5cfc6)

**Comment by tango4j** [comment]
/ok to test [3b3ec63](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/3b3ec6350c0a21988fdf746ace5abac163f0b8a3)

**Comment by tango4j** [comment]
@nithinraok 
All requested changes are addressed. Please review now that checks are passing.
`bypass_pre_encode` is staying because streaming Sortformer and other encoders need that feature.

File: examples/asr/conf/transformer/transformer_ctc_bpe.yaml
Comment (nithinraok) [inline examples/asr/conf/transformer/transformer_ctc_bpe.yaml:1]: can you rename this to `transformer_stacking_ctc_bpe.yaml` to align with other config file naming.

**Review by nithinraok — DISMISSED** [review]
Minor change requested. LGTM otherwise.

**Comment by ipmedenn** [comment]
Please don't merge until I check some feature

**Comment by tango4j** [comment]
/ok to test [f602751](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/f602751a99e8f56fcd0aa2edd88ebd7df49eb40d)

File: examples/asr/conf/transformer/transformer_ctc_bpe.yaml
Comment (tango4j) [reply examples/asr/conf/transformer/transformer_ctc_bpe.yaml:1]: Changed the file name and also changed the `name:` in the yaml to `name: "Transformer-Stacking-CTC-BPE"`

**Comment by tango4j** [comment]
/ok to test [ea5d3df](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/ea5d3dffda602c3ae93857e093468a4b6f40a8e4)

**Review by ipmedenn — APPROVED** [review]
LGTM!

**Comment by tango4j** [comment]
/ok to test [570c442](https://github.com/NVIDIA-NeMo/NeMo/pull/15703/commits/570c442f38c3f7bff131e6b3270610eeff31c551)
